sanoid

```diff
@@ -813,31 +821,42 @@ sub getsnaps {
 	my ($dev, $ino, $mode, $nlink, $uid, $gid, $rdev, $size, $atime, $mtime, $ctime, $blksize, $blocks) = stat($cache);
 
 	if ( $forcecacheupdate || ! -f $cache || (time() - $mtime) > $cacheTTL ) {
-		if (checklock('sanoid_cacheupdate')) {
-			writelock('sanoid_cacheupdate');
-			if ($args{'verbose'}) {
-				if ($args{'force-update'}) {
-					print "INFO: cache forcibly expired - updating from zfs list.\n";
-				} else {
-					print "INFO: cache expired - updating from zfs list.\n";
-				}
-			}
-			open FH, "$zfs get -Hrpt snapshot creation |";
-			@rawsnaps = <FH>;
-			close FH;
-
-			open FH, "> $cache" or die 'Could not write to $cache!\n';
-			print FH @rawsnaps;
-			close FH;
-			removelock('sanoid_cacheupdate');
-		} else {
-			if ($args{'verbose'}) { print "INFO: deferring cache update - valid cache update lock held by another sanoid process.\n"; }
+		if ( -f $cache && ! $forcecacheupdate && $no_need_for_cache_update ) {
```

Comment:
you still need the lock so reading of the cache file isn't done in the same moment as it's written by another process.

Reply:
Good point, you mean something like `if (checklock('sanoid_cacheupdate'))`? 

I propose another solution, which also solves your initial comment above:

How about instead of ignoring the age of the cache file (as originally discussed in the issue and PR), the allowed age of the cache file is non-persistently increased instead, and no changes to the `getsnaps` function are made? I.e., the default `cacheTTL` value of 900 (15 minutes) is overwritten with something larger such as 18000 (5 hours), if and only if a "monitor-*" command is supplied and no switch that would either create or prune snapshots is active?

The upside is that there are fewer code changes, fewer special cases, and even if some snapshots are deleted outside of sanoid (or the sanoid timer becomes inactive), after a couple hours the "--monitor-*" commands will let Nagios show that there is a problem (instead of erroneously showing that all is fine).

Reply:
checklock doesn't lock, you will need to put the cache reading between "writelock('sanoid_cacheupdate');" and "removelock('sanoid_cacheupdate');"

But looking at the existing code, reading the cache is also done without locks so there is already an issues which needs to be locked at (sanoid should only be running once not in parallel so this should only be an issue currently with monitor commands).

Your proposed solution sounds good.

Reply:
Then I'll make some code changes and commit again "soon".


Just one comment: I'm aware that checklock doesn't lock. But: The usecase that was the reason for the original issue cannot call "writelock"/"removelock", since the "--monitor-snapshots" etc commands are often run by the user "www-data" or "nagios" or something similar depending on Nagios/NRPE setup - Regardless of user, it is likely (and should be) a non-privileged account.

So unless sudo privileges are granted to the user calling "sanoid --monitor-*" by putting something in `/etc/sudoers.d`, or if that user is granted write access to `/run/sanoid` via `chmod`, then sanoid will fail for disk access rights reasons.

Reply:
@phreaker0 : I'm not used to pushing changes to git, but I think the final commit only changes the `cacheTTL` value in case of non-snapshot-modifying arguments. As discussed above.